src/main/kotlin/com/terraformation/backend/accelerator/db/CohortStore.kt

```diff
@@ -0,0 +1,121 @@
+import com.terraformation.backend.db.default_schema.tables.references.COHORTS
+import com.terraformation.backend.db.default_schema.tables.references.PARTICIPANTS
+import jakarta.inject.Named
+import java.time.InstantSource
+import org.jooq.Condition
+import org.jooq.DSLContext
+import org.jooq.impl.DSL
+import org.springframework.dao.DataIntegrityViolationException
+
+@Named
+class CohortStore(
+    private val clock: InstantSource,
+    private val dslContext: DSLContext,
+    private val cohortsDao: CohortsDao,
+) {
+  fun fetchOneById(cohortId: CohortId): ExistingCohortModel {
+    return fetch(COHORTS.ID.eq(cohortId)).firstOrNull() ?: throw CohortNotFoundException(cohortId)
+  }
+
+  fun findAll(): List<ExistingCohortModel> {
+    return fetch(null)
+  }
+
+  fun create(model: NewCohortModel): ExistingCohortModel {
+    requirePermissions { createCohort() }
+
+    val now = clock.instant()
+    val userId = currentUser().userId
+
+    val row =
+        CohortsRow(
+            createdBy = userId,
+            createdTime = now,
+            modifiedBy = userId,
+            modifiedTime = now,
+            name = model.name,
+        )
+
+    cohortsDao.insert(row)
+
+    return row.toModel()
+  }
+
+  fun delete(cohortId: CohortId) {
+    requirePermissions { deleteCohort(cohortId) }
+
+    try {
+      dslContext.transaction { _ ->
+        val rowsDeleted = dslContext.deleteFrom(COHORTS).where(COHORTS.ID.eq(cohortId)).execute()
+
+        if (rowsDeleted == 0) {
+          throw CohortNotFoundException(cohortId)
+        }
+      }
+    } catch (e: DataIntegrityViolationException) {
+      val hasParticipants =
+          dslContext.fetchExists(PARTICIPANTS, PARTICIPANTS.COHORT_ID.eq(cohortId))
+      if (hasParticipants) {
+        throw CohortHasParticipantsException(cohortId)
+      } else {
+        throw e
+      }
+    }
+  }
+
+  fun update(cohortId: CohortId, updateFunc: (ExistingCohortModel) -> ExistingCohortModel) {
+    requirePermissions { updateCohort(cohortId) }
+
+    val existing = fetchOneById(cohortId)
+    val updated = updateFunc(existing)
+
+    val rowsUpdated =
+        with(COHORTS) {
+          dslContext
+              .update(COHORTS)
+              .set(MODIFIED_BY, currentUser().userId)
+              .set(MODIFIED_TIME, clock.instant())
+              .set(NAME, updated.name)
+              .where(ID.eq(cohortId))
+              .execute()
+        }
+
+    if (rowsUpdated < 1) {
+      throw CohortNotFoundException(cohortId)
+    }
+  }
+
+  private fun fetch(condition: Condition?): List<ExistingCohortModel> {
+    val user = currentUser()
+
+    val participantIdsMultiset =
+        DSL.multiset(
+                DSL.select(PARTICIPANTS.ID)
+                    .from(PARTICIPANTS)
+                    .where(PARTICIPANTS.COHORT_ID.eq(COHORTS.ID))
+                    .orderBy(PARTICIPANTS.ID))
+            .convertFrom { result -> result.map { it[PARTICIPANTS.ID.asNonNullable()] } }
```

Comment:
Granted the data sets are likely to be pretty small so maybe it doesn't matter much, but I wonder if it's right to always fetch the list of participants. Will we actually use the participant IDs most of the time when we're working with cohorts, or will we only need them in specific contexts?

Reply:
So far I haven't seen any view that shows cohorts without participants. It seems they are always combined. Come to think of it, the only place cohorts are visible, you also see all the participants, projects, and managers associated. Do you think that should be implemented in another method (like `fetchWithFullParticipantData` or similar) or should it be baked into this? I am not sure that this method would be used on its own yet unless we were also going to merge the "base cohort data" (this fetch) and the "full participant / project data" together on the frontend.

Reply:
I was thinking things like drop-down menus of cohorts when editing participants. But if it's always needed, then doing it here is fine. If we find it's unnecessary we can make this work like `PlantingSiteStore` where the caller can say whether to return the child objects.

Reply:
I ended up implementing the `CohortDepth` in the same way as `PlantingSiteDepth`.